src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1569,6 +1569,352 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
+        // Case to handle 16 byte gather operation.
+        if (j > 0) {
+          // We enter here only if maximum integer vector size is less than 512
+          // bits.
+          if (int_vec_enc == Assembler::AVX_256bit) {
+            vpermq(xtmp2, xtmp2, shuf_mask[j], int_vec_enc);
+          } else {
+            assert(int_vec_enc == Assembler::AVX_128bit, "");
+            vpshufd(xtmp2, xtmp2, shuf_mask[j], vlen_enc);
+          }
+        }
+        vpor(dst, dst, xtmp2, vlen_enc);
+      }
+    }
+  } else {
+    assert(elem_bt == T_SHORT, "");
+    // Loop to gather 4(64bit), 8(128bit), 16(256bit) or 32(512bit) short values
+    // from memory into vector using integral gather instruction.
+    for (int i = 0, j = 0; i < lane_count_subwords; i += lane_count_ints, j++) {
+      vpxor(xtmp2, xtmp2, xtmp2, int_vec_enc);
+      kxnorwl(gmask, gmask, gmask);
+      vgather_subword_avx3(elem_bt, xtmp2, base, offset_vec, idx_base, i,
+                           idx_vec, xtmp1, xtmp3, gmask, int_vec_enc);
+      if (vlen_enc == Assembler::AVX_512bit) {
+        // Case to handle 32 byte gather operation.
+        assert(int_vec_enc == Assembler::AVX_512bit, "");
+        // Appropriately permute 256 bit lane holding 16 shorts accumulated
+        // using 512 bit integral gather operation.
+        if (j > 0) {
+          vinserti64x4(dst, dst, xtmp2, j);
+        } else {
+          evmovdquq(dst, xtmp2, vlen_enc);
+        }
+      } else if (vlen_enc == Assembler::AVX_256bit) {
+        // Case to handle 16 byte gather operation.
+        if (int_vec_enc == Assembler::AVX_512bit) {
+          // All 16 short values are loaded in one short by 512 bit integral
+          // gather.
+          vmovdqu(dst, xtmp2);
+        } else {
+          assert(int_vec_enc == Assembler::AVX_256bit, "");
+          // Permute 8 short values loaded using 256 bit integral gather.
+          vinserti32x4(dst, dst, xtmp2, j);
+        }
+      } else if (vlen_enc == Assembler::AVX_128bit) {
+        if (j > 0) {
+          // We enter here only if maximum integer vector size is less than 256
+          // bits.
+          assert(int_vec_enc == Assembler::AVX_128bit, "");
+          vpslldq(xtmp2, xtmp2, 8, vlen_enc);
+        }
+        vpor(dst, dst, xtmp2, vlen_enc);
+      }
+    }
+  }
+
+  if (mask != knoreg) {
+    if (elem_bt == T_BYTE) {
+      evmovdqub(dst, mask, dst, false, vlen_enc);
+    } else {
+      assert(elem_bt == T_SHORT, "");
+      evmovdquw(dst, mask, dst, false, vlen_enc);
+    }
+  }
+}
+
+void C2_MacroAssembler::vgather8b(BasicType elem_bt, XMMRegister dst,
+                                  Register base, Register idx_base,
+                                  Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      movl(rtmp, Address(idx_base, i * 4));
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
```

Comment:
Do I understand this right that you are basically doing this?
`dst[i*4 .. i*4 + 3] = load_8bytes(base + (idx_base + i * 4) * 2)`
But this does not look like a gather, rather like 4 adjacent loads that pack the data together into a single 8*4 byte vector.

If so, maybe you should either leave a comment, or even rename the method.

Reply:
Loop scans over integral index array and pick the work from computed address,  indexes could be non-contiguous.

Reply:
Maybe you could have comment lines that state this, similar like in the documentation?
`dst[i] = load(base + 2 * load(idx_base + i * 4))`
Or maybe:
`dst[i] = base[idx_base[i * 4] * 2]`